packages/odyssey-storybook/src/components/odyssey-labs/DataComponents/dataFunctions.ts

```diff
@@ -26,98 +19,100 @@ export const filterData = ({
-
-        // General filtering for other columns
-        return row[id as keyof Person]
-          ?.toString()
-          .toLowerCase()
-          .includes(value.toString().toLowerCase());
-      });
-    });
-  }
-
-  // Implement sorting
-  if (sort && sort.length > 0) {
-    filteredData.sort((a, b) => {
-      for (const { id, desc } of sort) {
-        const aValue = a[id as keyof Person];
-        const bValue = b[id as keyof Person];
-
-        if (aValue < bValue) return desc ? 1 : -1;
-        if (aValue > bValue) return desc ? -1 : 1;
-      }
+  const searchFiltered = search
+    ? data.filter((row) =>
+        Object.values(row).some((value) =>
+          value.toString().toLowerCase().includes(search.toLowerCase()),
+        ),
+      )
+    : data;
+
+  const columnFiltered = filters
+    ? searchFiltered.filter((row) =>
+        filters.every(({ id, value }) => {
+          if (value === null || value === undefined) {
+            return true;
+          }
+
+          if (Array.isArray(value)) {
+            return value.some((arrayValue) => {
+              const filterValue =
+                typeof arrayValue === "object" ? arrayValue.value : arrayValue;
+              return row[id as keyof Person]
+                ?.toString()
+                .toLowerCase()
+                .includes(filterValue.toString().toLowerCase());
+            });
+          }
+
+          if (id === "startLetter" && typeof row.name === "string") {
+            const firstLetter = row.name[0]?.toLowerCase();
+            if (value === "vowel") return "aeiou".includes(firstLetter);
+            if (value === "consonant") return !"aeiou".includes(firstLetter);
+            return true;
+          }
+
+          return row[id as keyof Person]
+            ?.toString()
+            .toLowerCase()
+            .includes(value.toString().toLowerCase());
+        }),
+      )
+    : searchFiltered;
+
+  const sorted =
+    sort && sort.length > 0
+      ? [...columnFiltered].sort((a, b) =>
+          sort.reduce((result, { id, desc }) => {
+            if (result !== 0) return result;
+
+            const aValue = a[id as keyof Person];
+            const bValue = b[id as keyof Person];
+
+            return desc
+              ? bValue < aValue
+                ? -1
+                : bValue > aValue
+                  ? 1
+                  : 0
+              : aValue < bValue
+                ? -1
+                : aValue > bValue
+                  ? 1
+                  : 0;
+          }, 0),
+        )
+      : columnFiltered;
 
-      return 0;
-    });
-  }
-
-  // Implement pagination
   const startRow = (page - 1) * resultsPerPage;
   const endRow = startRow + resultsPerPage;
-  filteredData = filteredData.slice(startRow, endRow);
 
-  return filteredData;
+  return sorted.slice(startRow, endRow);
 };
 
-export const reorderData = <T extends { id: string | number }>({
+export const reorderData = <Data extends { id: string | number }>({
```

Comment:
What is `Data` normally if not something that has a string or number `id`? Are we not setting this to `Data extends MRT_Columns` or something?